src/mo_frontend/bi_match.ml

```diff
@@ -192,10 +223,22 @@ let choose_under_constrained ctx er lb c ub =
     | t, _ ->
       ErrorUnderconstrained.add er lb c ub;
       if t = Non then ub else lb
-     
+    (* | _ ->
+     fail_under_constrained lb c ub *)
+
 let bi_match_typs ctx =
   let flexible c = ConSet.mem c ctx.var_set in
-  
+
+  let rec bi_match_list_result p rel eq inst any xs1 xs2 =
+    match (xs1, xs2) with
+    | x1::xs1', x2::xs2' ->
+      (match p rel eq inst any x1 x2 with
+      | Some inst -> bi_match_list_result p rel eq inst any xs1' xs2'
+      | None -> Result.Error (inst, ([x1], [x2])))
+    | [], [] -> Ok inst
+    | _, _ -> Result.Error (inst, (xs1, xs2))
```

Comment:
`List.combine` will trap when only one of these lists is nonempty.
If we just stop of first error then we don't need lists here, just a pair